[SYCL][CUDA][libclc] Add asynchronous barrier (#5303)
intel/llvm: sycl/doc/extensions/Barrier/Barrier.asciidoc

`void arrive_copy_async()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete. Before the arrive operation is triggered, the pending count on the barrier is increased by 1, so after the arrival there is no change to the pending count.

`void arrive_copy_async_no_inc()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete.

`void wait(arrival_token arrival)` : Executes wait operation, blocking until the predetermined number of work items reach arrive.

`void test_wait(arrival_token arrival)` : Checks whether all the arrivals have already happened for the current cycle, returning `true` if they did and `false` if `wait(arrival)` would block.

`static constexpr uint64_t max()` : Returns the maximum value of the expected and pending counts supported by the implementation. Exceeding this in the internal state of the barrier is undefined behavior.

==== Sample Header

[source, c++]
----
namespace sycl {
namespace ext {
namespace oneapi {

> class barrier {

Reviewer: does this really need to have that "reach" interface compared to std::barrier that it draws similarity with?

Author: I am not sure what do you mean by "reach interface".

Reviewer: I mean that [std::barrier](https://en.cppreference.com/w/cpp/thread/barrier) is happy with just arrive/wait/drop, and no token. Can we simplify this API (it's likely will add chances that other backends would be able to support it too)?

Author: While removing this token would make interface more similar to `std::barrier`, I disagree that doing so would make it easier to implement for other backends. If other backends do not need this token they can pass around a dummy value. Meanwhile implementing this without the token for CUDA would lead to more complicated implementation and most likely additional limitations, such as only one barrier being usable at once.

Reviewer: > If other backends do not need this token they can pass around a dummy value.

But the token is used as an input in the core `wait` API, it can't be dummy.

> Meanwhile implementing this without the token for CUDA would lead to more complicated implementation and most likely additional limitations, such as only one barrier being usable at once.

CUDA implementation can use "token" under the hood, of course.

I don't have any practical suggestions, just expressing a desire for a simpler and a more standard interface for the feature. I will rely on @Pennycook review/approval for the feature definition and then review the implementation.

Author: > But the token is used as an input in the core wait API, it can't be dummy.

I guess I was not clear enough. I meant that an  implementation that does need the token can return a dummy value from `arrive` and ignore that token in the `wait` call.

> CUDA implementation can use "token" under the hood, of course.

I have a feeling that hiding the token under the hood would, depending on how it is implemented, either limit the functionality, or be inefficient or complicated. Although I can't provide any concrete arguments from the top of my head. I need to think about this some more.

  [implementation defined internal state]

public:
  using arrival_token = [implementation defined];

  // barriers cannot be moved or copied
  barrier(const barrier &other) = delete;
  barrier(barrier &&other) noexcept = delete;
  barrier &operator=(const barrier &other) = delete;
  barrier &operator=(barrier &&other) noexcept = delete;

  void initialize(uint32_t expected_count);
  void invalidate();
  arrival_token arrive();
  arrival_token arrive_and_drop();
  arrival_token arrive_no_complete(int32_t count);
  arrival_token arrive_and_drop_no_complete(int32_t count);
  void arrive_copy_async();
  void arrive_copy_async_no_inc();
  void wait(arrival_token arrival);
  void test_wait(arrival_token arrival);

>   arrival_token arrive_and_wait();

Reviewer: this one is missing a documentation

Author: fixed

== Dependencies

This extension is written against the SYCL 2020 specification, Revision 4.

== Feature Test Macro

This extension provides a feature-test macro as described in the core SYCL
specification section 6.3.3 "Feature test macros".  Therefore, an
implementation supporting this extension must predefine the macro
`SYCL_EXT_ONEAPI_CUDA_ASYNC_BARRIER` to one of the values defined in the table
below. Applications can test for the existence of this macro to determine if
the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's APIs the implementation supports.

[%header,cols="1,5"]
|===
|Value |Description
|1     |Initial extension version.  Base features are supported.
|===

== Overview

This extension introduces asynchronous barrier for CUDA devices. This extends `group_barrier` by splitting it into into two calls - arrive and wait. The wait call blocks until the predetermined number of work items in the same work group call arrive. This is also very similar to https://en.cppreference.com/w/cpp/thread/barrier[`std::barrier`] introduced in c++20.

> Implementing this requires some space in local memory, where the state of the barrier is kept (this is true even for the CUDA compute capability 8 that support this functionality in hardware).

Reviewer 2: I'm not sure what the comment about CUDA compute capability 8 means here, exactly.  Are you saying that CUDA compute capability 8 requires local memory usage for the instruction(s) it provides, or that the extension uses local memory even where it's not required?

I think things might be clearer if you combine this paragraph with the one beneath it, something like this:

> This extension introduces a `barrier` class, encapsulating the state of an asynchronous barrier.  It must be used only in local memory, and reserves space in local memory even on hardware with native support (e.g. devices with CUDA Compute Capability 8.0).  Once a `barrier` is initialized, it can be reused for multiple cycles of arrivals and waits.

Author: done

This extension introduces `barrier` class. It must be used only in local memory. Once a `barrier` is initialized it can be reused for multiple cycles of arrivals and waits.

> This extension can only be used on devices that support independent forward progress.

Reviewer 2: ```suggestion
This extension can only be used on devices that support Independent Thread Scheduling (i.e. devices with CUDA Compute Capability 7.0 or higher).
```

=== Cycle of arrivals and waits

> When predetermined number of arrivals happen, the barrier moves into the next cycle. That unblocks the waits for current cycle. That is any wait that was previously called with the arrival token from this cycle stops blocking and future calls to wait with arrival token from this cycle will not block. The pending count is also reset and any future arrivals happen in the next cycle. However, at least one wait (or test wait returning `true`) must happen with arrival token from current cycle before any arrivals can happen in the next cycle. Another arrival happening before the wait is undefined behavior.

Reviewer 2: ```suggestion
After a predetermined number of arrivals, the barrier moves into the next cycle. Any `wait` that was previously called with the arrival token from this cycle is unblocked, and future calls to `wait` with the arrival token from this cycle will not block. The pending count is reset and any future arrivals happen in the next cycle. However, at least one `wait` (or `test_wait` returning `true`) must happen with the arrival token from the current cycle before any arrivals can happen in the next cycle. Another arrival happening before the `wait` is undefined behavior.
```

> Wait and test wait can only be called with an arrival token from current cycle (in which case wait will block until the barrier moves into next cycle) or the previous cycle (in which case wait will not block). Calling wait or test wait with an arrival token from any other cycle is undefined behavior.

Reviewer 2: ```suggestion
`wait` and `test_wait` can only be called with an arrival token from the current cycle (in which case `wait` will block until the barrier moves into next cycle) or the previous cycle (in which case `wait` will not block). Calling `wait` or `test_wait` with an arrival token from any other cycle results in undefined behavior.
```

=== Interface

> `barrier` class has the follwing member functions:

Reviewer 2: ```suggestion
`barrier` class has the following member functions:
```

> `void initialize(uint32_t expected_count)` : Initializes the barrier with given count of arrivals before wait unblocks. It only needs to be called by one work-item in work group. After the initialization a barrier operation (such as `group_barrier()`)needs to be executed by all work-items using the `barrier` object before they can use the newly initialized `barrier` object.

Reviewer 2: ```suggestion
`void initialize(uint32_t expected_count)` : Initializes the barrier with an expected number of arrivals, representing the number of arrivals required to unblock calls to `wait`. This function only needs to be called by one work-item in work group. After the initialization a barrier operation (such as `group_barrier()`) needs to be executed by all work-items using the `barrier` object before they can use the newly initialized `barrier` object. If `expected_count` is greater than the value returned by `max()`, behavior is undefined.
```

Reviewer 2: I left this at "only needs to be", but should it be stronger (i.e. "should only be")?  Is the behavior well-defined if multiple work-items initialize the barrier?  Is it unsafe to call `initialize` on a previously initialized `barrier` without calling `invalidate` first?

Author: PTX documentation is not very clear on the subject, but it does suggest that initializing multiple times is fine, while invalidating multiple times is not.

Alternatively we could be more strict and specify that initialize should also only be called once.

> `void invalidate()` : This must be called on an initialized barrier before the memory it is using can be repurposed for any other use. It only needs to be called by one work-item in work group. Before the invalidation a barrier operation (such as `group_barrier()`) needs to be executed by all work-items using the `barrier` object, after the last call on the `barrier`. After the invalidation a barrier operation (such as `group_barrier()`) needs to be executed by any work-items reusing the memory before the memory is reused. Calling any member function, except `initialize()` on an invalidated barrier is undefined behavior.

Reviewer 2: ```suggestion
`void invalidate()` : Invalidates a previously initialized barrier, enabling its associated memory to be repurposed. This function only needs to be called by one work-item in the work group. Before the invalidation a barrier operation (such as `group_barrier()`) needs to be executed by all work-items using the `barrier` object, after the last call on the `barrier`. After the invalidation a barrier operation (such as `group_barrier()`) needs to be executed by any work-items reusing the memory before the memory is reused. Calling any member function except `initialize()` on an invalidated barrier results in undefined behavior.
```

Reviewer 2: Again, should this be "should be called by one work-item"?  The last sentence implies that calling `invalidate` multiple times is not allowed.

> `arrival_token arrive()` : Executes arrival operation and returns a token that is needed for the wait call corresponding to this arrival.

Reviewer 2: ```suggestion
`arrival_token arrive()` : Executes arrival operation and returns a token that is needed for the `wait` call corresponding to this arrival.
```

> `arrival_token arrive_and_drop()` : Reduces expected arrival count for future cycles by one, executes arrival operation and returns a token that is needed for the wait call corresponding to this arrival.

Reviewer 2: ```suggestion
`arrival_token arrive_and_drop()` : Reduces expected arrival count for future cycles by one, executes arrival operation and returns a token that is needed for the `wait` call corresponding to this arrival.
```

> `arrival_token arrive_no_complete(int32_t count)` : Executes arrival operation that counts as `count` arrivals and returns a token that is needed for the wait call corresponding to this arrival. This must not be the last arrival that causes the cycle to complete - it would be undefined behavior.

Reviewer 2: ```suggestion
`arrival_token arrive_no_complete(int32_t count)` : Executes arrival operation that counts as `count` arrivals and returns a token that is needed for the `wait` call corresponding to this arrival. If this is the last arrival that causes the cycle to complete, behavior is undefined.
```

Reviewer 2: It's not clear from reading this why the function exists, why the behavior is undefined or how a developer can use this function safely.  Would it make sense to add a non-normative note explaining why the behavior is undefined?

Reviewer 2: It feels like there should be a restriction on `count` here given `max()` below.  I'm not sure if the requirement should be that the result of `count` arrivals cannot cause the total number of arrivals to exceed `expected_count` or that it cannot exceed `max()`.

Author: Clarified limitation on count and added some suggestions how this function can be used.

> `arrival_token arrive_and_drop_no_complete(int32_t count)` : Reduces expected arrival count for future cycles by one, executes arrival operation that counts as `count` arrivals and returns a token that is needed for the wait call corresponding to this arrival. This must not be the last arrival thet causes the cycle to complete - it would be undefined behavior.

Reviewer 2: ```suggestion
`arrival_token arrive_and_drop_no_complete(int32_t count)` : Reduces expected arrival count for future cycles by one, executes arrival operation that counts as `count` arrivals and returns a token that is needed for the `wait` call corresponding to this arrival. If this is the last arrival that causes the cycle to complete, behavior is undefined.
```

Reviewer 2: Same comment about `count` as above.

`void arrive_copy_async()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete. Before the arrive operation is triggered, the pending count on the barrier is increased by 1, so after the arrival there is no change to the pending count.

> `void arrive_copy_async_no_inc()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete.

Reviewer 2: I think adding some non-normative examples to the extension would help to clarify what these functions do.  From reading the description alone I'm having trouble working out what they do or how to use them.

Author: Added some examples.

> `void wait(arrival_token arrival)` : Executes wait operation, blocking until the predetermined number of work items reach arrive.

Reviewer 2: ```suggestion
`void wait(arrival_token arrival)` : Executes wait operation, blocking until the predetermined number of work items have called `arrive`.
```

`bool test_wait(arrival_token arrival)` : Checks whether all the arrivals have already happened for the current cycle, returning `true` if they did and `false` if `wait(arrival)` would block.

> `void arrive_and_wait()` : Equivalent to calling `this->wait(this->arrive())`.

Reviewer 2: ```suggestion
`void arrive_and_wait()` : Equivalent to calling `wait(this->arrive())`.
```

Author: Done, assuming you wanted removal of both `this->`.

> `static constexpr uint64_t max()` : Returns the maximum value of the expected and pending counts supported by the implementation. Exceeding this in the internal state of the barrier is undefined behavior.

Reviewer 2: I don't think users should be concerned with the internal state of the barrier.  How about:
```suggestion
`static constexpr uint64_t max()` : Returns the maximum value of the expected and pending counts supported by the implementation.
```
...with the note about undefined behavior moved to the other interfaces?

Author: done

> = SYCL_EXT_ONEAPI_BARRIER

Author: For now this extension is implemented only for CUDA. However, I don't think it is in general impossible to implement this for other backends. I have given it some thought and at least I have an idea how this could be implemented for lower sm versions that do not have hardware support for this. So I am not sure if we want to name this as a CUDA specific extension.

> = SYCL_EXT_ONEAPI_BARRIER

Author: Maybe I was unclear in the description of this PR. Now I clarified that this is not intended to be CUDA-only extension, just that for now we only have CUDA implementation.

> = SYCL_EXT_ONEAPI_BARRIER

Reviewer 2: I remain skeptical.  By tying this API to the number of work-items expected to arrive, you're permitting synchronization at a finer granularity than may be guaranteed to be supported by other hardware/backends.  For example, I can't think of a way to implement this extension on top of a generic OpenCL device, relying only on the execution model described in the OpenCL specification.

> = SYCL_EXT_ONEAPI_BARRIER

Author: That is true. When I was thinking about software implementation of barrier I figured I would need either independent forward progress or named barriers that can specify the number of participating work-items. Both of these features are available in CUDA, but not OpenCL. I also checked at HIP, and it looks like it supports nothing similar. 

So I think we need to determine if a common extension for CUDA, esimd and possibly level 0 (I did not look into that) makes sense.

> = SYCL_EXT_ONEAPI_BARRIER

Reviewer 2: Exactly.  And even a named barrier might not be sufficient.  SPIR-V exposes named barriers ([OpNamedBarrierInitialize](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpNamedBarrierInitialize) and [OpMemoryNamedBarrier](https://www.khronos.org/registry/SPIR-V/specs/unified1/SPIRV.html#OpMemoryNamedBarrier)), but these instructions are specified in terms of how many sub-groups they synchronize rather than how many work-items they synchronize.

I do think a common extension for split barriers makes sense.  But I wonder if it would make sense to start with something simpler (e.g. a barrier with exactly the same control-flow requirements and semantics as SYCL work-group barrier, but with separate `arrive` and `wait` phases).  I've not thought about this enough, but I suspect an API that allowed work-items to effectively ignore `arrive` and fallback to a regular barrier inside of `wait` would be portable and still permit acceleration when the right hardware is available. @FMarno is trying to do something like this already with work-group copies (see https://github.com/intel/llvm/pull/4950).

> = SYCL_EXT_ONEAPI_BARRIER

Author: > but these instructions are specified in terms of how many sub-groups they synchronize rather than how many work-items they synchronize.

I don't think that is an issue for a general use case with all work items in a sub-group either participating or not participating in a barrier. However, it does become an issue when participation in a barrier is not based on just sub-group. I think we can work around this issue by requiring that all subgroups are converged upon calling arrive and wait if independent forward progress is not supported. So I think we could have software implementation based on these two, but I don't see them implemented in libclc.

> I've not thought about this enough, but I suspect an API that allowed work-items to effectively ignore arrive and fallback to a regular barrier inside of wait would be portable and still permit acceleration when the right hardware is available.

Unfortunately that is not exactly the same functionality. Counterexample would be work items in a work group executing arrives for 2 barriers before waiting on these two barriers in different orders (that is possible even without independent forward progress if different orders occur in different sub-groups). This would work with true async barrier and deadlock with your suggestion.

> = SYCL_EXT_ONEAPI_BARRIER

Reviewer 2: > I think we can work around this issue by requiring that all subgroups are converged upon calling arrive and wait if independent forward progress is not supported.

Are you suggesting that all work-items in the sub-group would be required to encounter arrive/wait _and_ that the number of work-items expected would have to be a multiple of the sub-group size?

> Counterexample would be work items in a work group executing arrives for 2 barriers before waiting on these two barriers in different orders (that is possible even without independent forward progress if different orders occur in different sub-groups). This would work with true async barrier and deadlock with your suggestion.

You're right, I hadn't considered corner cases like this.  Have you got an example of what that code would look like?  I'm trying to imagine how likely this pattern is to occur, and how easy it would be to write it accidentally.

If what you said before was broadened to say that all work-items in a work-group must be converged upon calling `wait`, then I think this sort of re-ordering would naturally be disallowed.

> = SYCL_EXT_ONEAPI_BARRIER

Author: > Are you suggesting that all work-items in the sub-group would be required to encounter arrive/wait and that the number of work-items expected would have to be a multiple of the sub-group size?

Yes.

> If what you said before was broadened to say that all work-items in a work-group must be converged upon calling wait, then I think this sort of re-ordering would naturally be disallowed.

Yeah, if we add such restriction the trivial implementation is possible. However I am not sure if adding such a limitation makes the async barrier useful at all. With such limitation all use cases can be simplified to use regular group barrier, which would also be more efficient.

> = SYCL_EXT_ONEAPI_BARRIER

Reviewer 3: I think this could get icky very fast for "remainder" subgroups if the subgroup size doesn't evenly divide the work-group size.  If what we need is a subgroup barrier (or even a work-group barrier) it seems easier to describe that directly.

As @pennycook alluded to in comments above and below we have been looking at a very basic asynchronous barrier similar to `cm_sbarrier` that we believe can be supported on all hardware.  It will more or less follow the same convergence rules as the existing barrier but will allow arriving at the barrier separately from waiting at the barrier.  We believe this can provide a performance improvement over an ordinary combined barrier for hardware that has two different phases, but safely degrade to a combined barrier for hardware that does not.

> = SYCL_EXT_ONEAPI_BARRIER

Author: OK, I am getting convinced that we can not expose all or most of the features from CUDA in a general extension. So I will change this extension, as suggested in the first comment of this thread, to make it obvious that it is CUDA only.

:source-highlighter: coderay
:coderay-linenums-mode: table

// This section needs to be after the document title.
:doctype: book
:toc2:
:toc: left
:encoding: utf-8
:lang: en

:blank: pass:[ +]

// Set the default source code type in this document to C++,
// for syntax highlighting purposes.  This is needed because
// docbook uses c++ and html5 uses cpp.
:language: {basebackend@docbook:c++:cpp}

== Introduction
IMPORTANT: This specification is a draft.

NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc. used by permission by Khronos.

This document describes an extension that adds `barrier`, which acts similarly to https://en.cppreference.com/w/cpp/thread/barrier[`std::barrier`].

NOTE: CUDA can hardware accelerate from compute capability 8.

== Notice

Copyright (c) 2021 Intel Corporation.  All rights reserved.

== Status

Working Draft

> This is a preview extension specification, intended to provide early access to a feature for review and community feedback. When the feature matures, this specification may be released as a formal extension@.

Author: removed

Because the interfaces defined by this specification are not final and are subject to change they are not intended to be used by shipping software products.

== Version

Revision: 1

== Contact
Tadej Ciglarič, Codeplay (tadej 'dot' ciglaric 'at' codeplay 'dot' com)

== Dependencies

This extension is written against the SYCL 2020 specification, Revision 4.

== Feature Test Macro

This extension provides a feature-test macro as described in the core SYCL
specification section 6.3.3 "Feature test macros".  Therefore, an
implementation supporting this extension must predefine the macro

> `SYCL_EXT_ONEAPI_ASYNC_BARRIER` to one of the values defined in the table

Author: fixed

This extension introduces `barrier` class. It must be used only in local memory. Once a `barrier` is initialized it can be reused for multiple cycles of arrivals and waits.

=== Cycle of arrivals and waits

When predetermined number of arrivals happen, the barrier moves into the next cycle. That unblocks the waits for current cycle. That is any wait that was previously called with the arrival token from this cycle stops blocking and future calls to wait with arrival token from this cycle will not block. The pending count is also reset and any future arrivals happen in the next cycle. However, at least one wait (or test wait returning `true`) must happen with arrival token from current cycle before any arrivals can happen in the next cycle. Another arrival happening before the wait is undefined behavior.

Wait and test wait can only be called with an arrival token from current cycle (in which case wait will block until the barrier moves into next cycle) or the previous cycle (in which case wait will not block). Calling wait or test wait with an arrival token from any other cycle is undefined behavior.

=== Interface

`barrier` class has the follwing member functions:

> `void initialize(uint32_t expected_count)` : Initializes the barrier with given count of arrivals before wait unblocks. It only needs to be called by one thread in work group. After the initialization a barrier operation (such as `group_barrier()`)needs to be executed by all threads using the `barrier` object before they can use the newly initialized `barrier` object.

Author: Yes. Fixed.

> `void invalidate()` : This must be called on an initializted barrier before the memory it is using can be repurposed for any other use. It only needs to be called by one thread in work group. Before the invalidation a barrier operation (such as `group_barrier()`) needs to be executed by all threads using the `barrier` object, after the last call on the `barrier`. After the invalidation a barrier operation (such as `group_barrier()`) needs to be executed by any threads reusing the memory before the memory is reused. Calling any member function, except `initialize()` on an invalidated barrier is undefined behavior.

Author: fixed

`void arrive_copy_async()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete. Before the arrive operation is triggered, the pending count on the barrier is increased by 1, so after the arrival there is no change to the pending count.

`void arrive_copy_async_no_inc()` : Schedules arrive operation to be triggered asynchronously when all previous asynchronous memory copies initiated by the calling work item complete.

`void wait(arrival_token arrival)` : Executes wait operation, blocking until the predetermined number of work items reach arrive.

`void test_wait(arrival_token arrival)` : Checks whether all the arrivals have already happened for the current cycle, returning `true` if they did and `false` if `wait(arrival)` would block.

`static constexpr uint64_t max()` : Returns the maximum value of the expected and pending counts supported by the implementation. Exceeding this in the internal state of the barrier is undefined behavior.

==== Sample Header

[source, c++]
----
namespace sycl {
namespace ext {

> namespace oneapi {

Author: done

class barrier {
  [implementation defined internal state]

public:
  using arrival_token = [implementation defined];

  // barriers cannot be moved or copied
  barrier(const barrier &other) = delete;
  barrier(barrier &&other) noexcept = delete;
  barrier &operator=(const barrier &other) = delete;
  barrier &operator=(barrier &&other) noexcept = delete;

  void initialize(uint32_t expected_count);
  void invalidate();
  arrival_token arrive();
  arrival_token arrive_and_drop();
  arrival_token arrive_no_complete(int32_t count);
  arrival_token arrive_and_drop_no_complete(int32_t count);
  void arrive_copy_async();
  void arrive_copy_async_no_inc();
  void wait(arrival_token arrival);
  void test_wait(arrival_token arrival);
  arrival_token arrive_and_wait();
  static constexpr uint64_t max();
};

} // namespace oneapi
} // namespace ext
} // namespace sycl
----

== Issues

None.

//. asd
//+
//--
//*RESOLUTION*: Not resolved.
//--

> 

Author: removed

> . Is `barrier` the best name? Reasons for that name are that it is mostly in line with c+\+20 `std::barrier` and CUDA has the same name for this functionality. However it might be confusing with `group_barrier`, which is not present in c++20 and has a different name in CUDA - `__syncthreads`. Earlier version of CUDA docs called this `awbarrier`. Now that name is deprecated and they call it asynchronous barrier in text and `barrier` in code. Related PTX instructions use `mbarrier`.

Reviewer 2: `barrier` does seem to make sense here, given that its design is aligned with `std::barrier` and the [cuda::barrier](https://nvidia.github.io/libcudacxx/extended_api/synchronization_primitives/barrier.html) from libcu++.  Other names I've heard for this concept are "non-blocking barriers" and "split barriers".

Author: Added other ideas

--
*RESOLUTION*: Not resolved.
--

== Revision History

[cols="5,15,15,70"]
[grid="rows"]
[options="header"]
|========================================
|Rev|Date|Author|Changes
|1|2022-01-07|Tadej Ciglarič|*Initial public working draft*
|========================================

== Resources
* https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#aw-barrier
* https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#parallel-synchronization-and-communication-instructions-mbarrier
* https://nvidia.github.io/libcudacxx/extended_api/synchronization_primitives/barrier.html

//************************************************************************
//Other formatting suggestions:
//
//* Use *bold* text for host APIs, or [source] syntax highlighting.
//* Use +mono+ text for device APIs, or [source] syntax highlighting.
//* Use +mono+ text for extension names, types, or enum values.
//* Use _italics_ for parameters.

> //************************************************************************

Reviewer 2: You can remove this.

Author: removed

> = SYCL_EXT_ONEAPI_BARRIER

Reviewer 4: I guess we need to rename the document to `sycl_ext_oneapi_barrier.asciidoc` and move from `Barrier` to `supported` directory.
Tagging @gmlueck to confirm. 
I was hoping to find some guidance in https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc, but I was not able to find recommendations for the file naming.

Reviewer 5: There is an explanation of naming in the recently added [README-process.md](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/README-process.md) document.  The name of the file, the name of the extension, and the name of the feature-test macro must all agree.  Assuming you want to keep the existing feature-test macro, the name of the extension should be "sycl_ext_oneapi_cuda_async_barrier" and the name of the file should be "sycl_ext_oneapi_cuda_async_barrier.asciidoc".  The file should reside in either "supported", "proposed", or "experimental" according to its status.

Please also change the specification document to follow the specification template [template.asciidoc](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/template.asciidoc).  The format looks close now, but ther boilerplate of the first few sections should be changed.

Author: done

the implementation supports this feature, or applications can test the macro's
value to determine which of the extension's APIs the implementation supports.

[%header,cols="1,5"]
|===
|Value |Description
|1     |Initial extension version.  Base features are supported.
|===

== Overview

This extension introduces asynchronous barrier for CUDA devices. This extends `group_barrier` by splitting it into into two calls - arrive and wait. The wait call blocks until the predetermined number of work items in the same work group call arrive. This is also very similar to https://en.cppreference.com/w/cpp/thread/barrier[`std::barrier`] introduced in c++20.

Implementing this requires some space in local memory, where the state of the barrier is kept (this is true even for the CUDA compute capability 8 that support this functionality in hardware).

This extension introduces a `barrier` class, encapsulating the state of an asynchronous barrier. It must be used only in local memory, and reserves space in local memory even on hardware with native support (e.g. devices with CUDA Compute Capability 8.0). Once a barrier is initialized, it can be reused for multiple cycles of arrivals and waits.

> This extension can only be used on devices that support Independent Thread Scheduling (i.e. devices with CUDA Compute Capability 7.0 or higher) when using CUDA backend. On other backend the interface of `barrier` is not defined or declared. Attempting to call any functions defined in this extension with CUDA backend and a device with lower compute capability will cause kernel launch to fail.

Reviewer 5: > On other backend the interface of `barrier` is not defined or declared. Attempting to call any functions defined in this extension with CUDA backend and a device with lower compute capability will cause kernel launch to fail.

The part about "`barrier` is not defined or declared" does not  make sense to me.  We don't know at compile-time which backend the kernel will be submitted to.  Therefore, it seems like the `barrier` class always needs to be declared, so long as the implementation supports this extension.

I think it would be more correct to say something like:

> The `barrier` class is an optional kernel feature as defined by the core SYCL specification section 5.7 "Optional kernel features".  This class may only be used by kernels that are submitted to a device that has the CUDA backend.  Attempting to submit a kernel using this class to another device causes the kernel invocation command (e.g. `parallel_for`) to  throw a synchronous exception with the `errc::kernel_not_supported` error code.

That wording assumes this class can run on *any* CUDA device.  If instead it only works on certain CUDA devices, we should introduce a new aspect, which applications can use to query whether the CUDA device supports this class.  In that case, the wording should be more like:

> The `barrier` class is an optional kernel feature as defined by the core SYCL specification section 5.7 "Optional kernel features".  This class may only be used by kernels that are submitted to a device that has the `aspect::ext_oneapi_XXX` aspect.  Attempting to submit a kernel using this class to another device causes the kernel invocation command (e.g. `parallel_for`) to  throw a synchronous exception with the `errc::kernel_not_supported` error code.

Author: It only works on certain devices, so I added an aspect and changed the specification to what is suggested. However I am not sure how to implement throwing a synchronous exception on launching a kernel using unsupported feature. If use the same implementation for device not supporting it, building the kernel will fail. Same will happen if libclc does not implement the functions for unsupported backend. If I instead call `__builtin_trap()` the error will happen asynchronously when kernel executes that instructions.

Reviewer 5: The behavior with the synchronous exception will depend on the general [design for "optional device features"](https://github.com/intel/llvm/blob/sycl/sycl/doc/design/OptionalDeviceFeatures.md).  Once that design is in place, you will just need to decorate the `class barrier` definition with the appropriate C++ attribute.  That will avoid the compilation error and will cause the runtime to throw the synchronous exception.

Author: OK, I guess that means this extension needs to wait until that design is implemented. Or can we merge it as it is and add that attribute later?

Reviewer 5: You do not need to wait for the optional device features design to be implemented.  As you say, we can add the C++ attribute later once that design is in place.  In the mean time, your extension will have the following limitations:

* Applications that target heterogeneous devices will need to be careful using your extension if one of the devices does not support this CUDA feature.  They may need to split kernels using your extension into a separate translation unit or link their application with `-fsycl-device-code-split=per_kernel` in order to avoid false compilation errors.

* If an application mistakenly submits a kernel using your extension to a device that does not support it, they may not get a synchronous exception as stated in the specification.  (They may get an online compilation failure or other error instead.)

Note that these limitations are similar for any "optional device feature".

Author: Added explanation of these limitations.

Reviewer 5: Thanks for adding the explanation.  It would probably be better to add this in a "[NOTE]", though, which will emphasize that it is non-normative.

Author: Changed to a note.

Author: Thanks, done.

Reviewer 5: Thanks!

Implementing this requires some space in local memory, where the state of the barrier is kept (this is true even for the CUDA compute capability 8 that support this functionality in hardware).

> This extension introduces a `barrier`` class, encapsulating the state of an asynchronous barrier. It must be used only in local memory, and reserves space in local memory even on hardware with native support (e.g. devices with CUDA Compute Capability 8.0). Once a barrier is initialized, it can be reused for multiple cycles of arrivals and waits.

Reviewer 2: Looks like there's a duplicate backtick here (`barrier``)

Author: fixed
